File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        
        dummy = ListNode(0,head)

        left = dummy
        right = head

        i = n 
        while i != 0 :
            right = right.next
            i -=1
        
        while right :
            right = right.next
            left = left.next
        
        left.next = left.next.next
        return dummy.next

        # Here i am starting at left = 0 dummy ll 
        # then i am taking right pointer to n steps from starting head 
        # and then we iterate both left and right together until right goes to the other 
        # end such that are left is at now exactly n + 1 from right end and we then 
        # update left.next = left.next.next

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Map<Node, Node> mymap = new HashMap<>();

        mymap.put(null, null);

        Node curr = head;
        while (curr != null) {
            mymap.put(curr, new Node(curr.val));
            curr = curr.next;
        }

        curr = head;
        while (curr != null) {
            mymap.get(curr).next = mymap.get(curr.next);
            mymap.get(curr).random = mymap.get(curr.random);
            curr = curr.next;
        }

        return mymap.get(head);
    }
}
